### PR TITLE
export feature for tracks and booths is done

### LINKED ISSUE #1652: Export tracks and booths
It would be useful to have similar export options like we do for proposals.

### PR DESCRIPTION
**Pull request by namangupta01**
closes #1652 
- [X] Tracks
- [X] Booths

### CODE REVIEW

**Comment by namangupta01** [comment]
Here's hows it looks now : 
screen shot 2017-10-07 at 5 17 25 am
screen shot 2017-10-07 at 5 18 46 am

**Comment by namangupta01** [comment]
Exporting in PDF :
All tracks : 
screen shot 2017-10-07 at 5 20 02 am

Confirmed Tracks : 

screen shot 2017-10-07 at 5 20 17 am

**Comment by namangupta01** [comment]
Exporting All Tracks as CSV :
screen shot 2017-10-07 at 5 24 18 am

Exporting Confirmed Tracks as CSV: 
screen shot 2017-10-07 at 5 24 31 am

**Comment by namangupta01** [comment]
Exporting all tracks as XLS:
screen shot 2017-10-07 at 5 28 51 am

Exporting Confirmed Tracks as XLS:
screen shot 2017-10-07 at 5 29 10 am

**Comment by namangupta01** [comment]
For Booths:

screen shot 2017-10-08 at 2 01 30 pm

**Comment by namangupta01** [comment]
Exporting all booths as PDF :

screen shot 2017-10-08 at 2 01 41 pm


Exporting confirmed booths as PDF :

screen shot 2017-10-08 at 2 01 55 pm

**Comment by namangupta01** [comment]
Exporting all Booths as CSV : 
screen shot 2017-10-08 at 2 26 01 pm

Exporting confirmed Booths as CSV: 
screen shot 2017-10-08 at 2 25 51 pm

**Comment by namangupta01** [comment]
Exporting all Booths as XLS : 
screen shot 2017-10-08 at 2 04 39 pm

Exporting confirmed Booths as XLS:
screen shot 2017-10-08 at 2 05 36 pm

**Comment by differentreality** [comment]
@namangupta01 nice work! 
The buttons are supposed to be in a btn-group, but they are not showing up properly grouped. Please fix this. I guess you copy-pasted the code for the buttons from Events; the code has some small errors. Let's not repeat them to our new PR :)

**Comment by namangupta01** [comment]
@differentreality Thanks! What errors are you talking about? Travis failing error?

**Comment by differentreality** [comment]
No, the 3 buttons you added for exports are not grouped. Only the last 2 are. The first button is not grouped with the others. Look closely between 1st and 2nd button.

**Comment by namangupta01** [comment]
Is it now good to go?

screen shot 2017-10-08 at 6 23 29 pm
screen shot 2017-10-08 at 6 23 36 pm

**Comment by differentreality** [comment]
My point was to group them together, not separate them.

**Comment by namangupta01** [comment]
Even after adding all the buttons in the group space is still there:

screen shot 2017-10-08 at 7 38 18 pm
screen shot 2017-10-08 at 7 40 23 pm

**Comment by namangupta01** [comment]
Even after adding all the buttons in the group space is still there:

screen shot 2017-10-08 at 7 38 18 pm
screen shot 2017-10-08 at 7 40 23 pm

**Comment by namangupta01** [comment]
Even after adding all the buttons in the group, space is still there and other 2 dropdown (export csv and export xls ) are not working.

screen shot 2017-10-08 at 7 38 18 pm
screen shot 2017-10-08 at 7 40 23 pm

**Comment by namangupta01** [comment]
@differentreality Got it! 👍

**Comment by namangupta01** [comment]
It's done! All the buttons are grouped.
screen shot 2017-10-08 at 7 59 35 pm
screen shot 2017-10-08 at 7 59 28 pm

File: app/views/admin/tracks/tracks.xlsx.axlsx [hunk 1]
```diff
@@ -0,0 +1,6 @@
+wb = xlsx_package.workbook
+if @track_export_option == 'confirmed'
+  render partial: 'confirmed_tracks', locals: {:wb => wb}
```
Comment (differentreality) [inline app/views/admin/tracks/tracks.xlsx.axlsx]: Please add spaces within curly brackets, and use new ruby syntax

File: app/views/admin/booths/_all_booths.csv.haml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+             'Reasoning',
+             'Submitter Name',
+             'Submitter Relationship',
+             'Website Url',
+             'State']
+= CSV.generate_line ['All booths']
+= CSV.generate_line headers
+- @booths.each do |booth|
+  = CSV.generate_line(	[booth.id,
+  						booth.title,
```
Comment (differentreality) [inline app/views/admin/booths/_all_booths.csv.haml]: Indentation here is not OK. Please put everything vertically aligned, and use spaces for your indentation.

File: app/views/admin/tracks/_all_tracks.csv.haml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+             'Start Date',
+             'Start Date',
+             'Submitter Name',
+             'Included in Cfp',
+             'State']
+= CSV.generate_line ['All Tracks']
+= CSV.generate_line headers
+- @tracks.each do |track|
+  = CSV.generate_line(	[track.id,
+  track.name,
```
Comment (differentreality) [inline app/views/admin/tracks/_all_tracks.csv.haml]: Please fix your indentation on those lines. Same for booths csv export.

File: app/views/admin/tracks/_confirmed_tracks.xlsx.axlsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+          'Submitter Name',
+          'Included in Cfp',
+          'State']
+  sheet.add_row row, style: bold_style
+
+  @tracks.confirmed.each do |track|
+    row = []
+    row << track.id
+    row << track.name.to_s
+    row << track.description.to_s
```
[the remark this answers is not in the document]
Comment (differentreality) [reply app/views/admin/tracks/_confirmed_tracks.xlsx.axlsx]: @namangupta01 what about this comment?

[the remark this answers is not in the document]
Comment (namangupta01) [reply app/views/admin/tracks/_confirmed_tracks.xlsx.axlsx]: @differentreality i use wrap styles for the spreadsheet file, for pdf it automatically get adjusted.

**Comment by differentreality** [comment]
Your screenshots look very nice, however the reasoning field, realistically, will be much much larger than this. In that case this is a sample of how things will look like:

screenshot from 2017-10-13 15-42-10

**Comment by namangupta01** [comment]
@differentreality That's how its look like now:
screen shot 2017-10-13 at 10 02 04 pm

screen shot 2017-10-13 at 7 41 28 pm

**Comment by differentreality** [comment]
* Track PDF export is OK-ish. The title is too small though. As we discussed on IRC, you need to anticipate for realistic data. 4 letter words are not realistic titles of tracks.
* Track XLS does not wrap multiple lines in cells for me. Are you sure the screenshot you posted is the right one? (Without any manual changes?)

screenshot from 2017-10-17 14-39-06

* Track CSV is too wide, due to the title field occupying 80% of the view area. This is not helpful to the reader at all, specially since all data fit to my screen. 

screenshot from 2017-10-17 14-40-52

I'd expect to see something closer to this instead:
screenshot from 2017-10-17 14-43-34

There are similar issues with your Booth exports.
Please double check that your code works properly, before pinging for a review.

**Comment by namangupta01** [comment]
@differentreality i run the same downloaded file of xls for tracks and booths on three places :
1. **Mac** - It's showing properly wrapped text.
2. **Linux - LibreOfficeCalc -** It's not showing wrap property.
3. **Google Sheets** - Its showing properly wrapped just like it was showing mac.
I think there is some problem with **LibreOfficeCalc**, so please open the same xls file that you opened on **LibreOfficeCalc in Google Sheet** which will show correct format.(i.e it will show wrap_text property).

**Comment by namangupta01** [comment]
And please don't open csv file as xls file in LibreOfficeCalc, both have different formats. 
Actually csv file is not stored in rows and columns.

**Comment by differentreality** [comment]
Yes, you are right about the CSV. I did double check the XLS, and seems that wrapping does not work only for libre office. Even though I think we should still fix this, since we encountered it. OSEM is for open source conferences, it is safe to assume that people will be using open source tools, such as libreoffice.

Moreover, about XLS, I still think we should adapt the column widths. I am thinking more like this: 
`sheet.column_widths 10,20,35,35,20,35,20,10` So that we give a bit more room to title and url mainly.

**Comment by namangupta01** [comment]
@differentreality i think wrap_text style is also working in linux but it is not showing the property. I after going to the description column  did **alt + right arrow key** and the property shows up. Can you please have a look! It works for me!

**Comment by namangupta01** [comment]
But this is not a correct as alt + right arrow increase the width of column. But wrap_text style won't work without it in librecalc.

**Comment by namangupta01** [comment]
After giving more room to title and url in booths in xls and pdf, that's how it looks now :
PDF : 
screen shot 2017-10-22 at 11 08 14 pm

XLS: 
screen shot 2017-10-22 at 11 10 47 pm

**Comment by namangupta01** [comment]
After giving more room to title and room in tracks in xls and pdf, that's how it looks now :
PDF:
screen shot 2017-10-22 at 11 30 12 pm

XLS:
screen shot 2017-10-22 at 11 34 36 pm

XLS on google Sheets:
screen shot 2017-10-22 at 11 35 38 pm

**Comment by bear454** [comment]
@differentreality do those changes meet your expectations?

**Comment by differentreality** [comment]
@namangupta01 as we discussed text fields should have a bit more text in your examples so that it is realistic. In booths, submitter relationship field will definitely not be one word, as it is a text field. We can expect it to be _at least_ a few words, such as _community manager and long term contributor_
So I would give that a bit space as well.

Other than that, things look good, so feel free to squash and rebase.

**Comment by namangupta01** [comment]
@differentreality should i give some more space to submitter relationship column ?

**Comment by namangupta01** [comment]
Is this now good to go?

screen shot 2017-11-07 at 9 12 15 pm

screen shot 2017-11-07 at 9 13 45 pm

**Comment by namangupta01** [comment]
Is this now good to go?
screen shot 2017-11-07 at 9 13 45 pm
screen shot 2017-11-07 at 9 14 57 pm

**Comment by namangupta01** [comment]
Final commit related to the last screenshots is added and pr is rebased!

**Comment by differentreality** [comment]
@namangupta01 nice :+1:  what about squashing your commits so we can merge the PR?

**Comment by namangupta01** [comment]
@differentreality Okay! 👍

**Pull request merged**